// xpring-common-js/build/generated/xrp_ledger_grpc_pb.js
// GENERATED CODE -- DO NOT EDIT!

'use strict';
var grpc = require('grpc');
var account_info_pb = require('./account_info_pb.js');
var get_account_info_request_pb = require('./get_account_info_request_pb.js');
var get_fee_request_pb = require('./get_fee_request_pb.js');
var submit_signed_transaction_request_pb = require('./submit_signed_transaction_request_pb.js');
var submit_signed_transaction_response_pb = require('./submit_signed_transaction_response_pb.js');
var fee_pb = require('./fee_pb.js');

function serialize_io_xpring_AccountInfo(arg) {
  if (!(arg instanceof account_info_pb.AccountInfo)) {
    throw new Error('Expected argument of type io.xpring.AccountInfo');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_io_xpring_AccountInfo(buffer_arg) {
  return account_info_pb.AccountInfo.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_io_xpring_Fee(arg) {
  if (!(arg instanceof fee_pb.Fee)) {
    throw new Error('Expected argument of type io.xpring.Fee');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_io_xpring_Fee(buffer_arg) {
  return fee_pb.Fee.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_io_xpring_GetAccountInfoRequest(arg) {
  if (!(arg instanceof get_account_info_request_pb.GetAccountInfoRequest)) {
    throw new Error('Expected argument of type io.xpring.GetAccountInfoRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_io_xpring_GetAccountInfoRequest(buffer_arg) {
  return get_account_info_request_pb.GetAccountInfoRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_io_xpring_GetFeeRequest(arg) {
  if (!(arg instanceof get_fee_request_pb.GetFeeRequest)) {
    throw new Error('Expected argument of type io.xpring.GetFeeRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_io_xpring_GetFeeRequest(buffer_arg) {
  return get_fee_request_pb.GetFeeRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_io_xpring_SubmitSignedTransactionRequest(arg) {
  if (!(arg instanceof submit_signed_transaction_request_pb.SubmitSignedTransactionRequest)) {
    throw new Error('Expected argument of type io.xpring.SubmitSignedTransactionRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_io_xpring_SubmitSignedTransactionRequest(buffer_arg) {
  return submit_signed_transaction_request_pb.SubmitSignedTransactionRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_io_xpring_SubmitSignedTransactionResponse(arg) {
  if (!(arg instanceof submit_signed_transaction_response_pb.SubmitSignedTransactionResponse)) {
    throw new Error('Expected argument of type io.xpring.SubmitSignedTransactionResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_io_xpring_SubmitSignedTransactionResponse(buffer_arg) {
  return submit_signed_transaction_response_pb.SubmitSignedTransactionResponse.deserializeBinary(new Uint8Array(buffer_arg));
}


// RPCs available to interact with the XRP Ledger.
var XRPLedgerService = exports.XRPLedgerService = {
  // Get account info for an account on the XRP Ledger.
  getAccountInfo: {
    path: '/io.xpring.XRPLedger/GetAccountInfo',
    requestStream: false,
    responseStream: false,
    requestType: get_account_info_request_pb.GetAccountInfoRequest,
    responseType: account_info_pb.AccountInfo,
    requestSerialize: serialize_io_xpring_GetAccountInfoRequest,
    requestDeserialize: deserialize_io_xpring_GetAccountInfoRequest,
    responseSerialize: serialize_io_xpring_AccountInfo,
    responseDeserialize: deserialize_io_xpring_AccountInfo,
  },
  // Get the fee for a transaction on the XRP Ledger.
  getFee: {
    path: '/io.xpring.XRPLedger/GetFee',
    requestStream: false,
    responseStream: false,
    requestType: get_fee_request_pb.GetFeeRequest,
    responseType: fee_pb.Fee,
    requestSerialize: serialize_io_xpring_GetFeeRequest,
    requestDeserialize: deserialize_io_xpring_GetFeeRequest,
    responseSerialize: serialize_io_xpring_Fee,
    responseDeserialize: deserialize_io_xpring_Fee,
  },
  // Submit a signed transaction to the XRP Ledger.
  submitSignedTransaction: {
    path: '/io.xpring.XRPLedger/SubmitSignedTransaction',
    requestStream: false,
    responseStream: false,
    requestType: submit_signed_transaction_request_pb.SubmitSignedTransactionRequest,
    responseType: submit_signed_transaction_response_pb.SubmitSignedTransactionResponse,
    requestSerialize: serialize_io_xpring_SubmitSignedTransactionRequest,
    requestDeserialize: deserialize_io_xpring_SubmitSignedTransactionRequest,
    responseSerialize: serialize_io_xpring_SubmitSignedTransactionResponse,
    responseDeserialize: deserialize_io_xpring_SubmitSignedTransactionResponse,
  },
};

exports.XRPLedgerClient = grpc.makeGenericClientConstructor(XRPLedgerService);
