// xpring-common-js/build/src/serializer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const payment_pb_1 = require("../generated/payment_pb");
const transaction_pb_1 = require("../generated/transaction_pb");
const currency_pb_1 = require("../generated/currency_pb");
class Serializer {
    static transactionToJSON(transaction) {
        var object = transaction.toObject();
        this.convertPropertyName("account", "Account", object);
        this.convertPropertyName("sequence", "Sequence", object);
        this.convertPropertyName("signingPublicKeyHex", "SigningPubKey", object);
        const txFee = transaction.getFee();
        if (txFee == undefined) {
            return undefined;
        }
        object["Fee"] = this.xrpAmountToJSON(txFee);
        delete object.fee;
        delete object.payment;
        const transactionDataCase = transaction.getTransactionDataCase();
        switch (transactionDataCase) {
            case transaction_pb_1.Transaction.TransactionDataCase.PAYMENT: {
                const payment = transaction.getPayment();
                if (payment == undefined) {
                    return undefined;
                }
                Object.assign(object, this.paymentToJSON(payment));
                break;
            }
        }
        return object;
    }
    static paymentToJSON(payment) {
        const json = {
            TransactionType: "Payment",
            Destination: payment.getDestination(),
            Amount: {}
        };
        const amountCase = payment.getAmountCase();
        switch (amountCase) {
            case payment_pb_1.Payment.AmountCase.FIAT_AMOUNT: {
                const fiatAmount = payment.getFiatAmount();
                if (fiatAmount == undefined) {
                    return undefined;
                }
                const jsonFiatAmount = this.fiatAmountToJSON(fiatAmount);
                if (jsonFiatAmount == undefined) {
                    return undefined;
                }
                json.Amount = jsonFiatAmount;
                break;
            }
            case payment_pb_1.Payment.AmountCase.XRP_AMOUNT: {
                const xrpAmount = payment.getXrpAmount();
                if (xrpAmount == undefined) {
                    return undefined;
                }
                json.Amount = this.xrpAmountToJSON(xrpAmount);
                break;
            }
        }
        return json;
    }
    static fiatAmountToJSON(fiatAmount) {
        const json = fiatAmount.toObject();
        const currency = fiatAmount.getCurrency();
        if (currency == undefined) {
            return undefined;
        }
        json.currency = this.currencyToJSON(currency);
        return json;
    }
    static currencyToJSON(currency) {
        switch (currency) {
            case currency_pb_1.Currency.USD:
                return "USD";
        }
    }
    static xrpAmountToJSON(xrpAmount) {
        return xrpAmount.getDrops() + "";
    }
    static convertPropertyName(oldPropertyName, newPropertyName, object) {
        object[newPropertyName] = object[oldPropertyName];
        delete object[oldPropertyName];
    }
}
exports.default = Serializer;
